qt/widgets/common/src/AlgorithmHistoryWindow.cpp

```diff
@@ -642,16 +642,53 @@ void AlgHistoryTreeWidget::itemChecked(QTreeWidgetItem *item, int index) {
   std::vector<int> indicies;
   QModelIndex modelIndex;
 
-  do {
-    modelIndex = indexFromItem(item, index);
+  bool hadToCheck{false};
+  std::vector<QTreeWidgetItem *> itemsChecked;
+
+  // Iterates over the ancestors of the item the user checked and checks the
+  // item if it isn't checked already.
+  for (auto currentItem = item; currentItem;
+       currentItem = currentItem->parent()) {
+    modelIndex = indexFromItem(currentItem, index);
     indicies.emplace_back(modelIndex.row() + 1);
 
-    if (item->flags().testFlag(Qt::ItemIsUserCheckable)) {
-      item->setCheckState(index, Qt::Checked);
+    if (currentItem->flags().testFlag(Qt::ItemIsUserCheckable) &&
+        currentItem->checkState(index) != Qt::Checked) {
+      currentItem->setCheckState(index, Qt::Checked);
+      hadToCheck = true;
     }
 
-    item = item->parent();
-  } while (item);
+    // The item the user checked and any of the ancestors that needed to be
+    // checked here are added to a list so we know which item's children need to
+    // be copied.
+    if (hadToCheck || currentItem == item) {
+      itemsChecked.emplace_back(item);
+    }
+  }
+
+  // The HistoryView class holds a flat list of all the algorithms in the
+  // history, as well as the children of any algorithms that have been unrolled.
+  // This function passes to that class the indices of the algorithms that have
+  // been checked so that they can be unrolled. In order for the index of an
+  // algorithm in this class to be the same as the index of that algorithm in
+  // HistoryView, when an algorithm is unrolled a copy of its children is made,
+  // added to the tree after the algorithm, and hidden.
+  for (auto it = itemsChecked.rbegin(); it != itemsChecked.rend(); ++it) {
+    auto item{*it};
```

Comment:
Could this local variable have a different name, e.g. `currentItem`, to distinguish it from the input `item`?